src/assets/Azure.Core.Shared/OperationPoller.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Implementation of LRO polling logic.
+    /// </summary>
+    internal class OperationPoller
+    {
+        public delegate Response UpdateStatus(CancellationToken cancellationToken = default);
+        public delegate ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken);
+        public delegate bool HasCompleted();
+        public delegate Response GetRawResponse();
+        public delegate T Value<T>();
+
+        private DelayStrategy _delayStrategy;
+
+        public OperationPoller(DelayStrategy? fallbackStrategy = null)
+        {
+            _delayStrategy = new RetryAfterDelayStrategy(fallbackStrategy);
+        }
+
+        public virtual ValueTask<Response> WaitForCompletionResponseAsync(Operation operation, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
+            => WaitForCompletionResponseAsync(operation.UpdateStatusAsync, () => operation.HasCompleted, operation.GetRawResponse, suggestedInterval, cancellationToken);
+
+        public virtual async ValueTask<Response> WaitForCompletionResponseAsync(UpdateStatusAsync updateStatusAsync, HasCompleted hasCompleted, GetRawResponse getRawResponse, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await updateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (hasCompleted())
+                {
+                    return getRawResponse();
+                }
+
+                await Task.Delay(_delayStrategy.GetNextDelay(response, suggestedInterval), cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public virtual Response WaitForCompletionResponse(Operation operation, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
+            => WaitForCompletionResponse(operation.UpdateStatus, () => operation.HasCompleted, operation.GetRawResponse, suggestedInterval, cancellationToken);
+
+        public virtual Response WaitForCompletionResponse(UpdateStatus updateStatus, HasCompleted hasCompleted, GetRawResponse getRawResponse, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = updateStatus(cancellationToken);
+
+                if (hasCompleted())
+                {
+                    return getRawResponse();
+                }
+
+                Thread.Sleep(_delayStrategy.GetNextDelay(response, suggestedInterval));
```

Comment:
Can we use `cancellationToken.WaitHandle` here?

Reply:
Since this is the same behavior that was there before https://github.com/Azure/azure-sdk-for-net/pull/27017/files#diff-d98a5ed3ccb11d48b0a9d4261030c3b7d7487657ad6d54ec78c6d4984c2675e7L138 I think we should leave it as is and take that up in a future feature request.

Reply:
With 32 seconds delay, being able to cancel the wait becomes more important than with 1 second.  
So I'd like to have something like
```cs
cancellationToken.WaitHandle.WaitOne(_delayStrategy.GetNextDelay(response, suggestedInterval));
cancellationToken.ThrowIfCancellationRequested();
```

Reply:
It was already 32 second delays and more for the retry after header since that is what is returned in quite a few of our LROs just FYI.  I agree we should probably consider doing this as a feature enhancement in the future, and since its not an API change it should be easy to do at any time.